### PR TITLE
Add DialoGPT Agent

### PR DESCRIPTION
**Pull request by emilydinan**
**Patch description**
Add a DialoGPT agent. Inherits a lot from the existing GPT2Agent. Can try talking to the large model with 
```
parlai interactive -m hugging_face/dialogpt --add-special-tokens False --gpt2-size large --inference beam --beam-size 10 --beam-context-block-ngram 3 --beam-block-ngram 3 --beam-min-length 25
```

### CODE REVIEW

File: parlai/agents/hugging_face/README.md [hunk 1]
```diff
@@ -24,5 +24,26 @@ Enter Your Message: Parrots are
 parlai train_model -m hugging_face/gpt2 --add-special-tokens True --add-start-token True --gpt2-size medium -t convai2 -bs 2 -mf <modelfile>
 ```
 
-## Other models
-_Other models are coming soon -- stay tuned!_
\ No newline at end of file
+## DialoGPT
+
+To use GPT2, run your command with the flag: `-m hugging_face/dialogpt`.
```
Comment (klshuster) [inline parlai/agents/hugging_face/README.md:29]: nit: `To use DialoGPT`

File: parlai/agents/hugging_face/dialogpt.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    GPT2 is a multi-layer decoder-only Transformer.
+    The decoder is initialized with pretrained weights from Hugging Face.
+    Read more about this model here
+    <https://huggingface.co/transformers/model_doc/dialogpt.html>.
+
+    DialoGPT comes in three  sizes: small, medium, large.
+
+    If you are finetuning the Gpt2 agent as a dialogue agent, be sure
+    to run `--add-special-tokens True`. To examine the performance of the
```
Comment (klshuster) [inline parlai/agents/hugging_face/dialogpt.py:62]: perhaps to enforce this, we can override `train_step` to raise an error if this opt value is `False`? or even just set an attribute in `__init__` so we don't have to access opt every step

File: parlai/agents/hugging_face/dialogpt.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        agent = argparser.add_argument_group('DialoGPT Args')
+        agent.add_argument(
+            '--gpt2-size',
+            type=str,
+            default='small',
+            choices=['small', 'medium', 'large'],
+            help='Which size model to initialize.',
+        )
+        argparser.set_defaults(
+            delimeter='<|endoftext|>',
```
Comment (klshuster) [inline parlai/agents/hugging_face/dialogpt.py:78]: nit: think this should be `delimiter`

File: parlai/agents/hugging_face/hugging_face.py [hunk 4]
```diff
@@ -7,8 +7,9 @@
 """
 Integration with Hugging Face Transformers.
 
-Please see <https://huggingface.co/transformers/>. Currently, the only implementation is
-GPT2. To use this model, run with `-m hugging_face/gpt2`.
+Please see <https://huggingface.co/transformers/>. Currently, the only implementations
+iare GPT2 and DialoGPT. To use these models, run with `-m hugging_face/gpt2` or `-m
```
Comment (klshuster) [inline parlai/agents/hugging_face/hugging_face.py:11]: nit: `iare`

File: parlai/agents/hugging_face/README.md [hunk 5]
```diff
@@ -24,5 +24,26 @@ Enter Your Message: Parrots are
...
+Here is example output from the above command:
+```
+Enter Your Message: What do you think of parrots?
+[Gpt2]:  I love parrots. They are the best. I love them so much. I wish I had a pet parrot.
+```
+
+
+**Fine-tune DialoGPT medium on the ConvAI2 task:**
+```bash
+parlai train_model -m hugging_face/dialogpt --add-special-tokens True --delimiter '\n' --add-start-token True --gpt2-size medium -t convai2 -bs 2 -mf <modelfile>
```
Comment (klshuster) [inline parlai/agents/hugging_face/README.md:47]: @EricMichaelSmith didn't you have issues specifying  `--delimiter '\n'` on the command line?

File: parlai/agents/hugging_face/dialogpt.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            history_add_global_end_token='<|endoftext|>',
+            text_truncate=768,
```
Comment (klshuster) [inline parlai/agents/hugging_face/dialogpt.py:80]: not at all related to pr but this is an interesting truncation choice

File: parlai/agents/hugging_face/README.md [hunk 7]
```diff
@@ -24,5 +24,26 @@ Enter Your Message: Parrots are
...
+**Talk to DialoGPT large in interactive mode, with beam size 10, 3-gram beam blocking, and minimum beam length 25:**
+```bash
+parlai interactive -m hugging_face/dialogpt --add-special-tokens False --gpt2-size large --inference beam --beam-size 10 --beam-context-block-ngram 3 --beam-block-ngram 3 --beam-min-length 25
+```
+_Note:_ In the above command, we must have the flag `--add-special-tokens False` if we want to use the model _without_ finetuning it.
+
+Here is example output from the above command:
+```
+Enter Your Message: What do you think of parrots?
+[Gpt2]:  I love parrots. They are the best. I love them so much. I wish I had a pet parrot.
```
Comment (jxmsML) [inline parlai/agents/hugging_face/README.md:41]: [Dialogpt]:

File: parlai/agents/hugging_face/dialogpt.py (hunk 6 above)
Comment (emilydinan) [reply parlai/agents/hugging_face/dialogpt.py:80]: i can't remember why that's there...

File: parlai/agents/hugging_face/dialogpt.py (hunk 2 above)
Comment (emilydinan) [reply parlai/agents/hugging_face/dialogpt.py:62]: good idea

Comment (emilydinan) [reply parlai/agents/hugging_face/dialogpt.py:62]: there is actually a check in the init of the parent class, if bs > 1

**Comment by stephenroller** [comment]
Any chance at a test?

File: tests/nightly/gpu/test_dialogpt.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import unittest
+import parlai.utils.testing as testing_utils
+
+
+@testing_utils.skipUnlessGPU
+class TestDialogptModel(unittest.TestCase):
```
Comment (klshuster) [inline tests/nightly/gpu/test_dialogpt.py:12]: 😍

Comment (stephenroller) [reply tests/nightly/gpu/test_dialogpt.py:12]: :heart: :heart: :heart:

File: parlai/agents/hugging_face/README.md (hunk 5 above)
Comment (EricMichaelSmith) [reply parlai/agents/hugging_face/README.md:47]: > @EricMichaelSmith didn't you have issues specifying `--delimiter '\n'` on the command line?

Yeah @emilydinan I had to do `--delimiter $'\n'` to get the newline to be set correctly. See the first response to https://unix.stackexchange.com/questions/20035/how-to-add-newlines-into-variables-in-bash-script

**Pull request merged**